### PR TITLE
Helper methods / Simulator Creation

### PR DESCRIPTION
**Pull request by rafael-chavez**
maps locator strategy used in java-client to the locator strategy
Delegate certain commands to specific helper methods
Override locator strategies in specific commands.
Only create simulator if it does not exist in the list of current devices

### CODE REVIEW

**Comment by rafael-chavez** [comment]
@jlipps updated from this pull request. https://github.com/appium/appium-xcuitest-driver/pull/30

**Comment by rafael-chavez** [comment]
dependent pull request. https://github.com/appium/node-mobile-json-wire-protocol/pull/51

File: lib/commands/find.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    return await this.findElOrEls(cmd, '/element', _.isArray(args) ? args[0] : args)
+}
+
+commands.findElements = async function (cmd, ...args) {
+    return await this.findElOrEls(cmd, '/elements', _.isArray(args) ? args[0] : args)
+}
+
+helpers.findElOrEls = async function (cmd, endpoint, value) {
+    if (cmd === '-ios predicate string') {
+        //WebDriverAgent uses 'predicate string'
```
Comment (jlipps) [inline lib/commands/find.js]: `// Web...`

File: lib/commands/find.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import _ from 'lodash';
```
Comment (jlipps) [inline lib/commands/find.js]: this file should use 2-space indentation

File: lib/commands/gesture.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+
+let helpers = {}, extensions = {};
+
+helpers.mobileScroll = async function (opts={}) {
+  //proxy scroll request
```
Comment (jlipps) [inline lib/commands/gesture.js]: `// proxy`

File: lib/commands/index.js [hunk 4]
```diff
@@ -1,8 +1,15 @@
 import contextCommands from './context';
+import executeExtensions from './execute';
+import gestureExtensions from './gesture';
+import findExtensions from './find';
+import proxyHelperExtensions from './proxy-helper';
 
 let commands = {};
 
-for (let obj of [contextCommands]) {
+for (let obj of [
```
Comment (jlipps) [inline lib/commands/index.js]: would prefer simply:

```js
Object.assign(commands, executeExtensions, gestureExtensions, findExtensions,
              proxyHelperExtensions);
```

File: lib/commands/proxy-helper.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import { errors } from 'mobile-json-wire-protocol';
```
Comment (jlipps) [inline lib/commands/proxy-helper.js]: 2-space indents

File: lib/commands/proxy-helper.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+let helpers = {}, extensions = {};
+
+helpers.proxyCommand = async function (endpoint, method, body) {
+    if (!endpoint) {
+        throw new errors.BadParametersError('"WebDriverSpec Protocol Commands endpoint"');
```
Comment (jlipps) [inline lib/commands/proxy-helper.js]: s/commands/requires

**Comment by rafael-chavez** [comment]
dependent pull request https://github.com/appium/appium-base-driver/pull/50

**Comment by jlipps** [comment]
made some comments. mostly LGTM. @rafael-chavez should probably wait till https://github.com/appium/appium-xcuitest-driver/pull/39 is merged and rebase on top of that

**Comment by rafael-chavez** [comment]
@jlipps pulled in your changes and fixed styling. https://github.com/appium/appium-xcuitest-driver/pull/39

File: lib/driver.js [hunk 7]
```diff
@@ -156,6 +156,13 @@ class WebDriverAgentDriver extends BaseDriver {
           //TODO check for real device
           throw new errors.SessionNotCreatedError(`no simulator found with udid ${this.opts.udid}`);
       }
+    } else if (existingSim) {
```
Comment (jlipps) [inline lib/driver.js]: 2-space indentation

File: lib/webdriveragent.js [hunk 8]
```diff
@@ -111,15 +111,14 @@ class WebDriverAgent {
   createXcodeBuildSubProcess () {
 
     let args = [
-      '-workspace',
+      '-project',
       this.agentPath,
       '-scheme',
-      'XCTUITestRunner',
+      'WebDriverAgentRunner',
```
Comment (jlipps) [inline lib/webdriveragent.js]: what is this change for?

**Comment by jlipps** [comment]
looks good @rafael-chavez though i think rebase+force-push is better than merge. either way we don't want a merge commit in this PR so we'd need to squash.

File: lib/webdriveragent.js (hunk 8 above)
Comment (rafael-chavez) [inline lib/webdriveragent.js]: @jlipps this is the new scheme in the webdriveragent. the submodule pointer for WebDriverAgent is pointing to a relatively outdated webdriveragent commit.

Im proposing a merge [here]( https://github.com/appium/WebDriverAgent/pull/2/files#diff-354f30a63fb0907d4ad57269548329e3R14) any reason why we dont want to use the latest? it works.

**Comment by rafael-chavez** [comment]
k ill squash

File: lib/webdriveragent.js (hunk 8 above)
Comment (jlipps) [inline lib/webdriveragent.js]: nope, in fact i'm not sure why we run off of a fork @Jonahss?

File: lib/commands/find.js (hunk 2 above)
Comment (rafael-chavez) [inline lib/commands/find.js]: yes i updated to use 2 spaces.

File: lib/commands/proxy-helper.js (hunk 5 above)
Comment (rafael-chavez) [inline lib/commands/proxy-helper.js]: same here.

**Comment by rafael-chavez** [comment]
@jlipps rebased, forced update.

**Comment by jlipps** [comment]
@rafael-chavez lookin good. do you want to add a commit to this PR that updates the submodule reference for WDA?

**Comment by rafael-chavez** [comment]
sure thing @jlipps let me just get this pull request merged https://github.com/facebook/WebDriverAgent/pull/155 then ill set the pointer to that commit so that i dont have to make another pull request to point to this updated commit.

**Comment by jlipps** [comment]
👍

**Comment by jlipps** [comment]
@rafael-chavez i just merged fb's WDA into appium's WDA (which this project references). since it's a merge commit, can you update this to point to `b5b1c1e43ba9962402a8f0ced80b2fe55bbef60a` for the WDA submodule instead?

File: WebDriverAgent [hunk 9]
```diff
@@ -1 +1 @@
-Subproject commit bba1a925e7135058a30f645f54fcecd67e1ab9bf
+Subproject commit b5b1c1e43ba9962402a8f0ced80b2fe55bbef60a
```
Comment (rafael-chavez) [inline WebDriverAgent]: @jlipps pointing to appium WDA

File: lib/commands/execute.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import _ from 'lodash';
+import { util } from 'appium-support';
+import { errors } from 'appium-base-driver';
```
Comment (rafael-chavez) [inline lib/commands/execute.js]: we should be importing from base driver since thats where errors moved.

File: lib/commands/proxy-helper.js [hunk 11]
```diff
@@ -0,0 +1,16 @@
+import { errors } from 'appium-base-driver';
```
Comment (rafael-chavez) [inline lib/commands/proxy-helper.js]: same here.

File: lib/simulatorManagement.js [hunk 12]
```diff
@@ -20,6 +20,16 @@ async function createSim (caps, sessionId) {
   return await getSimulator(udid);
 }
 
+async function getExistingSim (deviceName, platformVersion) {
+  let devices = await getDevices(platformVersion);
+  for (var deviceIdx in devices) {
```
Comment (jlipps) [inline lib/simulatorManagement.js]: missed this before! this should probably be, to es2016 it and use our standards

```js
for (let device of _.values(devices)) {
  if (device.name === deviceName) {
    return await getSimulator(device.udid);
  }
} 
```

**Comment by jlipps** [comment]
@rafael-chavez just found one more thing! otherwise it's looking great

File: lib/simulatorManagement.js [hunk 13]
```diff
@@ -20,6 +21,16 @@ async function createSim (caps, sessionId) {
   return await getSimulator(udid);
 }
 
+async function getExistingSim (deviceName, platformVersion) {
+  let devices = await getDevices(platformVersion);
+  for (let device of _.values(devices)) {
```
Comment (rafael-chavez) [inline lib/simulatorManagement.js]: definitely looks cleaner than indexing, updated thanks @jlipps !

**Comment by jlipps** [comment]
ok merged! published as 1.0.3

**Pull request merged**